internal/plugin/service/serviceintegration/service_integration_data_source.go

```diff
@@ -0,0 +1,140 @@
+}
+
+func (s *serviceIntegrationDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "The Service Integration data source provides information about the existing Aiven Service Integration.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:   true,
+				Validators: []validator.String{endpointIDValidator},
+			},
+			"integration_id": schema.StringAttribute{
+				Description: "Service Integration Id at aiven",
+				Computed:    true,
+			},
+			"destination_endpoint_id": schema.StringAttribute{
+				Description: "Destination endpoint for the integration (if any)",
+				Computed:    true,
+				Validators:  []validator.String{endpointIDValidator},
+			},
+			"destination_service_name": schema.StringAttribute{
+				Description: "Destination service for the integration (if any)",
+				Required:    true,
+			},
+			"integration_type": schema.StringAttribute{
+				Description: "Type of the service integration. Possible values: " + schemautil.JoinQuoted(integrationTypes, ", ", "`"),
+				Required:    true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(integrationTypes...),
+				},
+			},
+			"project": schema.StringAttribute{
+				Description: "Project the integration belongs to",
+				Required:    true,
+			},
+			"source_endpoint_id": schema.StringAttribute{
+				Description: "Source endpoint for the integration (if any)",
+				Computed:    true,
+				Validators:  []validator.String{endpointIDValidator},
+			},
+			"source_service_name": schema.StringAttribute{
+				Description: "Source service for the integration (if any)",
+				Required:    true,
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"clickhouse_kafka_user_config":                clickhousekafka.NewDataSourceSchema(),
+			"clickhouse_postgresql_user_config":           clickhousepostgresql.NewDataSourceSchema(),
+			"datadog_user_config":                         datadog.NewDataSourceSchema(),
+			"external_aws_cloudwatch_metrics_user_config": externalawscloudwatchmetrics.NewDataSourceSchema(),
+			"kafka_connect_user_config":                   kafkaconnect.NewDataSourceSchema(),
+			"kafka_logs_user_config":                      kafkalogs.NewDataSourceSchema(),
+			"kafka_mirrormaker_user_config":               kafkamirrormaker.NewDataSourceSchema(),
+			"logs_user_config":                            logs.NewDataSourceSchema(),
+			"metrics_user_config":                         metrics.NewDataSourceSchema(),
+		},
+	}
+}
+
+// Read reads datasource
+// All functions adapted for resourceModel, so we use it as donor
+// Copies state from datasource to resource, then back, when things are done
+func (s *serviceIntegrationDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var o dataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &o)...)
```

Comment:
use `util.PlanStateToModel`

Reply:
That's useless function. It does nothing but makes you read it and match variable names to understand what it does.

Reply:
it saves the amount of lines that is necessary to implement a `Read` method, and it also has a comprehensive comment on it, so one wouldn't need to jump in it to understand its usage, if you believe that the naming is not good, feel free to rename it in this PR, otherwise I see no reason why shouldn't it be reused